urwid/event_loop/main_loop.py

```diff
@@ -242,66 +244,70 @@ def remove_alarm(self, handle) -> bool:
-        :meth:`watch_pipe`. You are responsible for closing the write end of
-        the pipe.
-
-        Returns ``True`` if the watch pipe exists, ``False`` otherwise
-        """
-        try:
-            watch_handle, pipe_rd = self._watch_pipes.pop(write_fd)
-        except KeyError:
-            return False
+    if not IS_WINDOWS:
+
+        def watch_pipe(self, callback: Callable[[bytes], bool]) -> int:
+            """
+            Create a pipe for use by a subprocess or thread to trigger a callback
+            in the process/thread running the main loop.
+
+            :param callback: function taking one parameter to call from within
+                             the process/thread running the main loop
+            :type callback: callable
+
+            This method returns a file descriptor attached to the write end of a
+            pipe. The read end of the pipe is added to the list of files
+            :attr:`event_loop` is watching. When data is written to the pipe the
+            callback function will be called and passed a single value containing
+            data read from the pipe.
+
+            This method may be used any time you want to update widgets from
+            another thread or subprocess.
+
+            Data may be written to the returned file descriptor with
+            ``os.write(fd, data)``. Ensure that data is less than 512 bytes (or 4K
+            on Linux) so that the callback will be triggered just once with the
+            complete value of data passed in.
+
+            If the callback returns ``False`` then the watch will be removed from
+            :attr:`event_loop` and the read end of the pipe will be closed. You
+            are responsible for closing the write end of the pipe with
+            ``os.close(fd)``.
+            """
+            import fcntl
+
+            pipe_rd, pipe_wr = os.pipe()
+            fcntl.fcntl(pipe_rd, fcntl.F_SETFL, os.O_NONBLOCK)
+            watch_handle = None
+
+            def cb() -> None:
+                data = os.read(pipe_rd, PIPE_BUFFER_READ_SIZE)
+                rval = callback(data)
+                if rval is False:
+                    self.event_loop.remove_watch_file(watch_handle)
+                    os.close(pipe_rd)
+
+            watch_handle = self.event_loop.watch_file(pipe_rd, cb)
+            self._watch_pipes[pipe_wr] = (watch_handle, pipe_rd)
+            return pipe_wr
+
+        def remove_watch_pipe(self, write_fd: int) -> bool:
+            """
+            Close the read end of the pipe and remove the watch created by
+            :meth:`watch_pipe`. You are responsible for closing the write end of
+            the pipe.
+
+            Returns ``True`` if the watch pipe exists, ``False`` otherwise
+            """
+            try:
+                watch_handle, pipe_rd = self._watch_pipes.pop(write_fd)
+            except KeyError:
+                return False
```

Comment:
This is mainly used to trigger a screen update from a signal (like the terminal resize) or from another process/thread. On windows we need something like this, maybe a socket pair connected over loopback?

Reply:
https://github.com/urwid/urwid/blob/master/urwid/_win32_raw_display.py#L247 ->
https://github.com/urwid/urwid/blob/master/urwid/_win32_raw_display.py#L165C77-L165C94 ->
https://github.com/urwid/urwid/blob/master/urwid/_raw_display_base.py#L99

And `socket` instance is used for sending `b"R"` to trigger update.
Unfortunately it's not enough, so issue #700 created and in process of debugging, when I have physical access to the Windows machine.